include/dxc/Support/dxcapi.extval.h

```diff
@@ -1,8 +1,160 @@
 #include "dxc/Support/dxcapi.use.h"
+#include "dxc/WinAdapter.h"
+#include "llvm/Support/raw_ostream.h"
+
 #include <cassert>
 #include <string>
 
 namespace dxc {
+
+class ExternalValidationHelper3 : public IDxcCompiler3 {
+
+public:
+  CComPtr<IDxcCompiler3> m_pCompiler3;
+  CComPtr<IDxcValidator> m_pValidator;
+
+  ExternalValidationHelper3() {
+    m_pCompiler3 = nullptr;
+    m_pValidator = nullptr;
+  }
+
+  HRESULT STDMETHODCALLTYPE Compile(
+      _In_ const DxcBuffer *pSource, ///< Source text to compile.
+      _In_opt_count_(argCount)
+          LPCWSTR *pArguments, ///< Array of pointers to arguments.
+      _In_ UINT32 argCount,    ///< Number of arguments.
+      _In_opt_ IDxcIncludeHandler
+          *pIncludeHandler,  ///< user-provided interface to handle include
+                             ///< directives (optional).
+      _In_ REFIID riid,      ///< Interface ID for the result.
+      _Out_ LPVOID *ppResult ///< IDxcResult: status, buffer, and errors.
+      ) override;
+
+  HRESULT STDMETHODCALLTYPE Disassemble(
+      _In_ const DxcBuffer
+          *pObject,     ///< Program to disassemble: dxil container or bitcode.
+      _In_ REFIID riid, ///< Interface ID for the result.
+      _Out_ LPVOID *ppResult)
+      override ///< IDxcResult: status, disassembly text, and errors.
+  {
+    return m_pCompiler3->Disassemble(pObject, riid, ppResult);
+  }
+
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid,
+                                           void **ppvObject) override {
+    return m_pCompiler3->QueryInterface(riid, ppvObject);
+  }
```

Comment:
This ExternalValidationHelper3 wrapper thingy is only ever going to be used by our internal test code I hope?  If not, then we have problems because this wrapper behaves in some suprising ways.

If it is only used by our internal test code it might still be worth putting some guard rails in place.  For example, do our tests actually call QueryInterface?  If they don't, then we should maybe make this just `std::abort`.

eg:

```c++
IDxcCompiler3* foo = // get one that's actually a ExternalValidationHelper3`

IDxcCompiler3* bar;
foo->QueryInterface(IID_PPV_ARGS(&bar));

// Sets bar to something other than foo, which is surprising.

foo->Validate() // uses m_pValidator
bar->Validate() // uses m_pCompiler3
```

Reply:
ExternalValidationHelper3's implementation of QueryInterface just calls through to m_pCompiler3's QueryInterface.  So the thing you get back is m_pCompiler3.